src/Nethermind/Nethermind.Blockchain/Blocks/BlockhashStore.cs

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText:2023 Demerzel Solutions Limited
+// SPDX-License-Identifier:LGPL-3.0-only
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.State;
+
+[assembly: InternalsVisibleTo("Nethermind.Blockchain.Test")]
+[assembly: InternalsVisibleTo("Nethermind.Merge.Plugin.Test")]
+namespace Nethermind.Blockchain.Blocks;
+
+public class BlockhashStore(IBlockFinder blockFinder, ISpecProvider specProvider, IWorldState worldState)
+    : IBlockhashStore
+{
+    private static readonly byte[] EmptyBytes = [0];
+
+    public void ApplyHistoryBlockHashes(BlockHeader blockHeader)
+    {
+        IReleaseSpec spec = specProvider.GetSpec(blockHeader);
+        if (!spec.IsEip2935Enabled || blockHeader.IsGenesis || blockHeader.ParentHash is null) return;
+
+        Address? eip2935Account = spec.Eip2935ContractAddress ?? Eip2935Constants.BlockHashHistoryAddress;
+        if (!worldState.AccountExists(eip2935Account)) return;
+
+        // TODO: find a better way to handle this - no need to have this check everytime
+        //      this would just be true on the fork block
+        BlockHeader parentHeader = blockFinder.FindParentHeader(blockHeader, BlockTreeLookupOptions.None);
+        if (parentHeader is not null && parentHeader!.Timestamp < spec.Eip2935TransitionTimestamp)
+            InitHistoryOnForkBlock(blockHeader, eip2935Account);
+        else
+            AddParentBlockHashToState(blockHeader, eip2935Account);
+    }
+
+    public Hash256? GetBlockHashFromState(BlockHeader currentHeader, long requiredBlockNumber)
+    {
+        IReleaseSpec? spec = specProvider.GetSpec(currentHeader);
+        if (requiredBlockNumber >= currentHeader.Number ||
+            requiredBlockNumber + Eip2935Constants.RingBufferSize < currentHeader.Number)
+        {
+            return null;
+        }
+        var blockIndex = new UInt256((ulong)(requiredBlockNumber % Eip2935Constants.RingBufferSize));
+        Address? eip2935Account = spec.Eip2935ContractAddress ?? Eip2935Constants.BlockHashHistoryAddress;
+        StorageCell blockHashStoreCell = new(eip2935Account, blockIndex);
+        ReadOnlySpan<byte> data = worldState.Get(blockHashStoreCell);
+        return data.SequenceEqual(EmptyBytes) ? null : new Hash256(data);
+    }
+
+    private void InitHistoryOnForkBlock(BlockHeader currentBlock, Address eip2935Account)
+    {
+        long current = currentBlock.Number;
+        BlockHeader header = currentBlock;
+        for (var i = 0; i < Math.Min(Eip2935Constants.RingBufferSize, current); i++)
+        {
+            // an extra check - don't think it is needed
```

Comment:
I think it is needed when you start devnet from genesis or from early blocks with 2935

Reply:
I think it shouldn't be needed if for loop header would take it into account and I think it does - this could use a test and removing this.

Reply:
so it is handled by the loop - I will remove this check.

genesis is added to the state in current implementation also, we have a test for this. for adding genesis we dont need to iterate to block 0, we need to iterate to block 1, because we add the parent hash to the state.